Add JSON escape strategy

c960657 opened the pull request:
The `js` escape strategy used `\xNN`-style escape sequences. This is not allowed in JSON, only `\uNNNN` is allowed.

This PR adds a new escape strategy, `json` that is similar to `js` except that it does not use `\xNN` but uses `\uNNNN` instead.

I know that I can use `json_decode()`, but it came as a surprise to me that `escape('js')` did not work. You probably shouldn't be generating JSON structures in your templates, but sometimes it comes in handy. My use-case was generating a piece of [JSON-LD](https://json-ld.org).

A different approach that would perhaps be less confusing to users would be to change `js` to not use `\xNN`. This output is still valid JavaScript, it just uses two more bytes per occurrence. Let me know what you think.

fabpot commented:
1.x is accepting new features, so it should be done on the 1.x branch instead.

xabbuh commented:
And since this is a new feature it should be part of a minor version increment in the changelog (not part of a patch/bugfix release).

c960657 commented:
Changed branch and moved to minor version.

The Fabbot failures are not in code I touched. The Travis failures on PHP 5.3 is a [separate issue](https://github.com/twigphp/Twig/pull/2582).

fabpot commented:
If that makes no difference, I would change the JS strategy instead of adding a new one, that would be less confusing. Also, if you rebase, tests would be ok and fabbot happy. Thanks.

c960657 committed:
Make js filter generate valid JSON

c960657 commented:
I have now modified the existing `js` filter to generate valid JSON.

The old approach is here: https://github.com/twigphp/Twig/compare/1.x...c960657:escape-json-bak

fabpot commented on test/Twig/Tests/Fixtures/tags/autoescape/strategy.legacy.test:
```diff
@@ -1,4 +1,4 @@
---TEST--
+  --TEST--
```
This change looks wrong to me.

fabpot reviewed (changes requested):
@xabbuh  @stof Do you think this is acceptable? Having longer encoded strings? I think it's not an issue, but want to double-check with you :)

c960657 committed:
Fix stray whitespace

c960657 commented:
If longer strings is an issue, we could use the shorter escape sequences for `\n`, `\t` etc.

fabpot commented:
@c960657 We need to balance output length and performance. If using `\n`/`\t` does not come with a performance penalty, go for it.

c960657 committed:
Use short escape sequences when applicable

c960657 commented:
I don't think there is any significant performance penalty.

I have added short codes for escape sequences supported in both JavaScript and JSON (JavaScript supports e.g. `\v` or `\.`, but JSON does not).

Also, inspired by [drupal_json_encode](https://api.drupal.org/api/drupal/includes%21common.inc/function/drupal_json_encode/7.x), `"` is still encoded using `\uHHHH` sequence.

fabpot commented on lib/Twig/Extension/Core.php:
```diff
@@ -1152,9 +1152,23 @@ function _twig_escape_js_callback($matches)
+    /*
+     * A few characters have short escape sequences in JSON and JavaScript.
+     * Escape sequences supported only by JavaScript, not JSON, are ommitted.
+     * \" is also supported but omitted, because the resulting string is not HTML safe.
```
Do you mean `\"` or `"` here?

c960657 replied:
Both, sort of :-)  I mean that the escape sequence for the character `"` is `\"`. Neither the raw character nor the escaped string is HTML safe.

fabpot commented:
Thank you @c960657.
